go-controller/pkg/node/iptables/iptables_manager.go

```diff
@@ -0,0 +1,402 @@
+	}
+	wantedRuleArgs, alreadyExists := c.chainRules[newChain]
+	if !alreadyExists {
+		c.chainRules[newChain] = newRuleArgs
+	} else {
+		for _, newRuleArg := range newRuleArgs {
+			alreadyExists = false
+
+			for _, existingRuleArg := range wantedRuleArgs {
+				if existingRuleArg.equal(newRuleArg) {
+					alreadyExists = true
+					break
+				}
+			}
+			if !alreadyExists {
+				c.chainRules[newChain] = append(c.chainRules[newChain], newRuleArg)
+			}
+		}
+	}
+	return c.reconcile()
+}
+
+// GetIPv4ChainRuleArgs must be passed a regex that matches one IPv4 or IPv6 address which is added to a map as a key
+// Chain will be selected with the associated regex match
+func (c *Controller) GetIPv4ChainRuleArgs(table iptables.Table, chain iptables.Chain) ([]RuleArgs, error) {
+	return getChainRuleArgs(c.iptV4, table, chain)
+}
+
+// GetIPv6ChainRuleArgs must be passed a regex that matches one IPv4 or IPv6 address which is added to a map as a key
+// Chain will be selected with the associated regex match
+func (c *Controller) GetIPv6ChainRuleArgs(table iptables.Table, chain iptables.Chain) ([]RuleArgs, error) {
+	return getChainRuleArgs(c.iptV6, table, chain)
+}
+
+// GetIPv4ChainRuleArgs must be passed a regex that matches one IPv4 or IPv6 address which is added to a map as a key
+// Chain will be selected with the associated regex match
+func getChainRuleArgs(ipt iptables.Interface, table iptables.Table, chain iptables.Chain) ([]RuleArgs, error) {
+	buf := bytes.NewBuffer(nil)
+
+	if err := ipt.SaveInto(table, buf); err != nil {
+		return nil, fmt.Errorf("failed to retrieve iptables table %s: %v", table, err)
+	}
+	rules := make([]RuleArgs, 0)
+	for _, line := range strings.Split(buf.String(), "\n") {
+		if strings.HasPrefix(line, fmt.Sprintf("-A %s", string(chain))) {
+			rules = append(rules, RuleArgs{
+				// cleave off the -A ${chain_name}
+				Args: strings.Split(line, " ")[2:],
+			})
+		}
+
+	}
+	return rules, nil
+}
+
+func processRules(ipt iptables.Interface, ownedChains []Chain, wantedChainRules map[Chain][]RuleArgs, existingChainRules map[Chain][]RuleArgs) error {
+	var err error
+	for _, wantedChain := range ownedChains {
+		err = execIPTablesWithRetry(func() error {
+			_, err = ipt.EnsureChain(wantedChain.Table, wantedChain.Chain)
+			if err != nil {
+				return fmt.Errorf("failed to ensure chain %s in table %s: %v", wantedChain.Chain, wantedChain.Table, err)
+			}
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("failed to process iptables rules in table %s: %v", wantedChain.Table, err)
+		}
+	}
+	for existingChain, existingRuleArgs := range existingChainRules {
+		if existingChain.Proto != ipt.Protocol() {
+			continue
+		}
+		wantedRuleArgs, found := wantedChainRules[existingChain]
+		if !found {
+			// we don't delete chains we don't manage, but we ensure rules are removed
+			continue
+		}
+		err = execIPTablesWithRetry(func() error {
+			_, err = ipt.EnsureChain(existingChain.Table, existingChain.Chain)
```

Comment:
Does this mean we will try to apply all of the iptables rules we manage every time we call `processRules`?

Reply:
yes

Reply:
Added a note to improve this later. Inspect state before blindly applying.